### PR TITLE
Liuliu/log-graph

### PR DESCRIPTION
**Pull request by liuliu-dev**
command: `dolt log --graph`

Steps for Drawing the Commit Graph
1. Calculate the Positions of Each Commit:

- Row Position: Determined by the commit's topological order, initialized using the index in the list. Adjust the vertical spacing between commits based on the commit message's height in `expandGraph` function.

- Column Position, calculated based on the positions of child commits, implemented in `computeColumnEnds` function:
  - No Children: Places the commit in a new column (furthest right) if it has no child commits, indicating the start of a new branch.
  - Branch Children: If a commit has branch children, it is positioned in the same column as its leftmost branch child.
  - Merge Children: For commits with merge children, diagonal lines ('\') are used to connect to these children, positioning the commit in an available column starting from its rightmost child.

2. Render the Graph:
Place a `*` at each commit's calculated position to mark it.
Connect commits using vertical (`|`) and diagonal (`\` or `/`) lines to represent branches and merges.

An example of the graph of [us-jails](https://www.dolthub.com/repositories/dolthub/us-jails/data/main):

Screenshot 2024-08-02 at 10 11 12 AM

### CODE REVIEW

**Comment by coffeegoddd** [comment]
@liuliu-dev __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>af879e6</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>af879e6</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@liuliu-dev __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>c23f150</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>c23f150</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@liuliu-dev __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>16c683e</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>16c683e</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@liuliu-dev __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>5695942</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>5695942</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by timsehn** [comment]
A couple of output graphs in code blocks would be cool to see.

**Review by timsehn — COMMENTED** [review]
Probably needs a couple bats tests.

**Comment by coffeegoddd** [comment]
@liuliu-dev __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>5ce04ab</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>5ce04ab</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@liuliu-dev __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>c924c8c</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>c924c8c</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@liuliu-dev __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>abacfb2</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>abacfb2</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

File: go/cmd/dolt/commands/log_graph.go [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+import (
+	"fmt"
+	"math"
+	"strings"
+
+	"github.com/dolthub/dolt/go/cmd/dolt/cli"
+	"github.com/dolthub/dolt/go/libraries/utils/argparser"
+	"github.com/dolthub/dolt/go/store/util/outputpager"
+)
+
```
Comment (macneale4) [inline go/cmd/dolt/commands/log_graph.go:26]: Can you summarize at the top here how the display works at a high level? Specifically, what information is necessary to render each commit, and how that types into the structures you've defined.

File: go/cmd/dolt/commands/log_graph.go [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+		}
+		if currentLine != "" {
+			wrappedLines = append(wrappedLines, currentLine)
+			totalRows++
+		}
+	}
+	return wrappedLines
+}
+
+func min(values ...int) int {
```
Comment (macneale4) [inline go/cmd/dolt/commands/log_graph.go:101]: There is a buildin function by the same name which takes only two args. Rename this to `minVal` to avoid confusion.

File: go/cmd/dolt/commands/log_graph.go [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+		if i < len(commits)-1 && commits[i+1].Commit.commitHash == commit.Commit.parentHashes[0] {
+			maxDistanceFromParent = math.Max(math.Abs(float64(commits[i+1].X-commit.X)), maxDistanceFromParent)
+		}
+
+		commitInfoHeight := getHeightOfCommit(commit)
+		posY += int(math.Max(float64(commitInfoHeight+1), maxDistanceFromParent))
+	}
+}
+
+func logGraph(pager *outputpager.Pager, apr *argparser.ArgParseResults, commitInfos []CommitInfo) {
```
Comment (macneale4) [inline go/cmd/dolt/commands/log_graph.go:337]: looks like the apr instance is only used to pass through the decoration flag. It would be better to use a more simple type.

File: go/cmd/dolt/commands/log_graph.go [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+
+	"github.com/dolthub/dolt/go/cmd/dolt/cli"
+	"github.com/dolthub/dolt/go/libraries/utils/argparser"
+	"github.com/dolthub/dolt/go/store/util/outputpager"
+)
+
+type CommitInfoWithChildren struct {
+	Commit           CommitInfo
+	Children         []string
+	X                int
```
Comment (macneale4) [inline go/cmd/dolt/commands/log_graph.go:30]: X/Y can get confusing based on your frame of reference. These are columns and rows of the terminal, correct? renaming to reflect that would help.

File: go/cmd/dolt/commands/log_graph.go [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+			// the y coordinate of the commit is initialized to the index of the commit as the commits are sorted
+			Y: y,
+		})
+	}
+
+	return commitsWithChildren
+}
+
+// wrap the commit message in a constrained width to better align the commit message with the graph
+func wrapTextOnWidth(text string, width int) []string {
```
Comment (macneale4) [inline go/cmd/dolt/commands/log_graph.go:73]: Awkward function name. `indentWithWrap` would be more straight forward

Comment (macneale4) [reply go/cmd/dolt/commands/log_graph.go:73]: Is wrapping word necessary? Does git do this? My suspicion is that it doesn't. The pager program will allow for right scrolling of the user wants it. I'm inclined to simplify the code. Indenting seems necessary, I just don't think that wrapping the words is required.

File: go/cmd/dolt/commands/log_graph.go [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+var branchColors = []string{
+	"\033[31m", // Red
+	"\033[32m", // Green
+	"\033[34m", // Blue
+	"\033[35m", // Magenta
+	"\033[36m", // Cyan
+	"\033[39m", // Default
+}
+
+type BranchPathType struct {
```
Comment (macneale4) [inline go/cmd/dolt/commands/log_graph.go:44]: Shorten to `BranchPath` - and document what it's used to track.

**Review by macneale4 — CHANGES_REQUESTED** [review]
I'm realizing now that I need a higher level description of the algorithm in order to properly review this. I want you to be overly heavy on documentation of fields and functions. Comments _in_ functions are almost always a sign that the function is too complex, and should be broken into multiple functions. Not always, just a rule of thumb. I'm seeing some pretty long functions in this code, and the descriptions of what they each do doesn't come close to the amount of code in the function.

**Comment by coffeegoddd** [comment]
@liuliu-dev __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>a0a8dab</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>a0a8dab</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@liuliu-dev __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>a5f41cc</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>a5f41cc</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

File: go/cmd/dolt/commands/log_graph.go [hunk 7]
```diff
@@ -0,0 +1,536 @@
...
+  Author: liuliu <liu@dolthub.com>
+  Date: Thu Aug 01 10:20:44 -0700 2024
+
+       Initialize data repository
+
+* ------------
+*
+*/
+
+type CommitInfoWithChildren struct {
```
Comment (macneale4) [inline go/cmd/dolt/commands/log_graph.go:162]: This type is only used in the package, so make it start with a lower case letter (makes it package private)

File: go/cmd/dolt/commands/log_graph.go [hunk 8]
```diff
@@ -0,0 +1,536 @@
...
+var branchColors = []string{
+	"\033[31m", // Red
+	"\033[32m", // Green
+	"\033[34m", // Blue
+	"\033[35m", // Magenta
+	"\033[36m", // Cyan
+	"\033[39m", // Default
+}
+
+type BranchPath struct {
```
Comment (macneale4) [inline go/cmd/dolt/commands/log_graph.go:179]: Lower case as well.

File: go/cmd/dolt/commands/log_graph.go (hunk 3 above)
Comment (liuliu-dev) [reply go/cmd/dolt/commands/log_graph.go]: I'm thinking adding more support in the next PR, such as `dolt log --graph --oneline`.

**Comment by coffeegoddd** [comment]
@liuliu-dev __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>e08f120</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>e08f120</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@liuliu-dev __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>154d90c</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>154d90c</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@liuliu-dev __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>2382bb3</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>2382bb3</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@coffeegoddd __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>eed2184</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>eed2184</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Review by macneale4 — APPROVED** [review]
```
     .  o ..                  
     o . o o.o                
          ...oo               
            __[]__            
         __|_o_o_o\__         
         \""""""""""/         
          \. ..  . /          
     ^^^^^^^^^^^^^^^^^^^^
```
